telemetry/record.go

```diff
@@ -225,6 +231,28 @@ func transformLocation(message *protos.Payload) {
 	}
 }
 
+// transformDetailedChargeState does a best-effort attempt to convert the DetailedChargeState field to a proper ChargingState Emum
+func transformDetailedChargeState(message *protos.Payload) {
+	for _, datum := range message.Data {
+		if datum.GetKey() == protos.Field_DetailedChargeState {
+			if strVal := datum.GetValue().GetStringValue(); strVal != "" {
+				datum.Value = &protos.Value{Value: &protos.Value_ChargingValue{ChargingValue: parseDetailedChargeState(strVal)}}
+			}
+			// There can be only one Field_DetailedChargeState Datum in the proto; abort once we've seen it.
+			return
+		}
+	}
+}
+
+// parseDetailedChargeState parses a location string (such as "Disconnected") into a *proto.Location type.
```

Comment:
The comment is slightly incorrect, "parses a location", should be "parses a charge state string"....